# Cancer Detection 

In [2]:
import pandas as pd

In [3]:
pwd

'/Users/pratik'

#### Breast Cancer Dataset

In [894]:
input_file = ("/Users/pratik/Desktop/Harrisburg University programs/Courses/Late Fall Courses 2022/ANLY 530 Principles of Machine Learning/Assignments & PPT's/Assignment 3/Datasets/wisc_bc_data.csv")
data = pd.read_csv(input_file)

In [895]:
data

id diagnosis  radius_mean  texture_mean  perimeter_mean  \
0     87139402         B        12.32         12.39           78.85   
1      8910251         B        10.60         18.95           69.28   
2       905520         B        11.04         16.83           70.92   
3       868871         B        11.28         13.39           73.00   
4      9012568         B        15.19         13.21           97.65   
..         ...       ...          ...           ...             ...   
564  911320502         B        13.17         18.22           84.28   
565     898677         B        10.26         14.71           66.20   
566     873885         M        15.28         22.41           98.92   
567     911201         B        14.53         13.98           93.86   
568    9012795         M        21.37         15.10          141.30   

     area_mean  smoothness_mean  compactness_mean  concavity_mean  \
0        464.1          0.10280           0.06981         0.03987   
1        346.4          0.09688           0.11470         0.06387   
2        373.2          0.10770           0.07804         0.03046   
3        384.8          0.11640           0.11360         0.04635   
4        711.8          0.07963           0.06934         0.03393   
..         ...              ...               ...             ...   
564      537.3          0.07466           0.05994         0.04859   
565      321.6          0.09882           0.09159         0.03581   
566      710.6          0.09057           0.10520         0.05375   
567      644.2          0.10990           0.09242         0.06895   
568     1386.0          0.10010           0.15150         0.19320   

     points_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0        0.03700  ...         13.50          15.64            86.97   
1        0.02642  ...         11.88          22.94            78.28   
2        0.02480  ...         12.41          26.44            79.93   
3        0.04796  ...         11.92          15.77            76.53   
4        0.02657  ...         16.20          15.73           104.50   
..           ...  ...           ...            ...              ...   
564      0.02870  ...         14.90          23.89            95.10   
565      0.02037  ...         10.88          19.48            70.89   
566      0.03263  ...         17.80          28.03           113.80   
567      0.06495  ...         15.80          16.93           103.10   
568      0.12550  ...         22.69          21.84           152.10   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0         549.1            0.1385             0.1266          0.12420   
1         424.8            0.1213             0.2515          0.19160   
2         471.4            0.1369             0.1482          0.10670   
3         434.0            0.1367             0.1822          0.08669   
4         819.1            0.1126             0.1737          0.13620   
..          ...               ...                ...              ...   
564       687.6            0.1282             0.1965          0.18760   
565       357.1            0.1360             0.1636          0.07162   
566       973.1            0.1301             0.3299          0.36300   
567       749.9            0.1347             0.1478          0.13730   
568      1535.0            0.1192             0.2840          0.40240   

     points_worst  symmetry_worst  dimension_worst  
0         0.09391          0.2827          0.06771  
1         0.07926          0.2940          0.07587  
2         0.07431          0.2998          0.07881  
3         0.08611          0.2102          0.06784  
4         0.08178          0.2487          0.06766  
..            ...             ...              ...  
564       0.10450          0.2235          0.06925  
565       0.04074          0.2434          0.08488  
566       0.12260          0.3175          0.09772  
567       0.10690          0.2606          0.07810  
568       0.19660

In [896]:
data = data.drop(columns = ["id"], axis = 1)

In [897]:
data.describe()

radius_mean  texture_mean  perimeter_mean    area_mean  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  points_mean  \
count       569.000000        569.000000      569.000000   569.000000   
mean          0.096360          0.104341        0.088799     0.048919   
std           0.014064          0.052813        0.079720     0.038803   
min           0.052630          0.019380        0.000000     0.000000   
25%           0.086370          0.064920        0.029560     0.020310   
50%           0.095870          0.092630        0.061540     0.033500   
75%           0.105300          0.130400        0.130700     0.074000   
max           0.163400          0.345400        0.426800     0.201200   

       symmetry_mean  dimension_mean  ...  radius_worst  texture_worst  \
count     569.000000      569.000000  ...    569.000000     569.000000   
mean        0.181162        0.062798  ...     16.269190      25.677223   
std         0.027414        0.007060  ...      4.833242       6.146258   
min         0.106000        0.049960  ...      7.930000      12.020000   
25%         0.161900        0.057700  ...     13.010000      21.080000   
50%         0.179200        0.061540  ...     14.970000      25.410000   
75%         0.195700        0.066120  ...     18.790000      29.720000   
max         0.304000        0.097440  ...     36.040000      49.540000   

       perimeter_worst   area_worst  smoothness_worst  compactness_worst  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       concavity_worst  points_worst  symmetry_worst  dimension_worst  
count       569.000000    569.000000      569.000000       569.000000  
mean          0.272188      0.114606        0.290076         0.083946  
std           0.208624      0.065732        0.061867         0.018061  
min           0.000000      0.000000        0.156500         0.055040  
25%           0.114500      0.064930        0.250400         0.071460  
50%           0.226700      0.099930        0.282200         0.080040  
75%           0.382900      0.161400        0.317900         0.092080  
max           1.252000      0.291000        0.663800         0.207500  

[8 rows x 30 columns]

In [898]:
list(data)

['diagnosis',
 'radius_mean',
 'texture_mean',
 'perimeter_mean',
 'area_mean',
 'smoothness_mean',
 'compactness_mean',
 'concavity_mean',
 'points_mean',
 'symmetry_mean',
 'dimension_mean',
 'radius_se',
 'texture_se',
 'perimeter_se',
 'area_se',
 'smoothness_se',
 'compactness_se',
 'concavity_se',
 'points_se',
 'symmetry_se',
 'dimension_se',
 'radius_worst',
 'texture_worst',
 'perimeter_worst',
 'area_worst',
 'smoothness_worst',
 'compactness_worst',
 'concavity_worst',
 'points_worst',
 'symmetry_worst',
 'dimension_worst']

#### Finding Levels of Tumor, which are Malignant and which are Benign. Checking the Diagnosis column and factorizing it for Benign and Malignant.

In [899]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   diagnosis          569 non-null    object 
 1   radius_mean        569 non-null    float64
 2   texture_mean       569 non-null    float64
 3   perimeter_mean     569 non-null    float64
 4   area_mean          569 non-null    float64
 5   smoothness_mean    569 non-null    float64
 6   compactness_mean   569 non-null    float64
 7   concavity_mean     569 non-null    float64
 8   points_mean        569 non-null    float64
 9   symmetry_mean      569 non-null    float64
 10  dimension_mean     569 non-null    float64
 11  radius_se          569 non-null    float64
 12  texture_se         569 non-null    float64
 13  perimeter_se       569 non-null    float64
 14  area_se            569 non-null    float64
 15  smoothness_se      569 non-null    float64
 16  compactness_se     569 non

In [914]:
list(data)

['diagnosis',
 'radius_mean',
 'texture_mean',
 'perimeter_mean',
 'area_mean',
 'smoothness_mean',
 'compactness_mean',
 'concavity_mean',
 'points_mean',
 'symmetry_mean',
 'dimension_mean',
 'radius_se',
 'texture_se',
 'perimeter_se',
 'area_se',
 'smoothness_se',
 'compactness_se',
 'concavity_se',
 'points_se',
 'symmetry_se',
 'dimension_se',
 'radius_worst',
 'texture_worst',
 'perimeter_worst',
 'area_worst',
 'smoothness_worst',
 'compactness_worst',
 'concavity_worst',
 'points_worst',
 'symmetry_worst',
 'dimension_worst']

#### Scaling the data.

In [912]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaled = scaler.fit_transform(data.drop(columns=["diagnosis"]))
print(scaled)

[[0.25268588 0.0906324  0.24227766 ... 0.32271478 0.24876799 0.08310376]
 [0.17128118 0.31247886 0.17614539 ... 0.27237113 0.27104278 0.136626  ]
 [0.19210564 0.24078458 0.18747841 ... 0.25536082 0.28247585 0.15590975]
 ...
 [0.3927777  0.42948935 0.38096883 ... 0.42130584 0.31736645 0.27994228]
 [0.35728146 0.14440311 0.34600235 ... 0.36735395 0.20520402 0.15125279]
 [0.68100715 0.18227934 0.67383042 ... 0.67560137 0.22964715 0.20739866]]


#### Creating dataframe with scaled data

In [915]:
data_new = pd.DataFrame(scaled, columns = ['radius_mean',
 'texture_mean',
 'perimeter_mean',
 'area_mean',
 'smoothness_mean',
 'compactness_mean',
 'concavity_mean',
 'points_mean',
 'symmetry_mean',
 'dimension_mean',
 'radius_se',
 'texture_se',
 'perimeter_se',
 'area_se',
 'smoothness_se',
 'compactness_se',
 'concavity_se',
 'points_se',
 'symmetry_se',
 'dimension_se',
 'radius_worst',
 'texture_worst',
 'perimeter_worst',
 'area_worst',
 'smoothness_worst',
 'compactness_worst',
 'concavity_worst',
 'points_worst',
 'symmetry_worst',
 'dimension_worst'])

In [916]:
data_new.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0     0.252686      0.090632        0.242278   0.135992         0.452920   
1     0.171281      0.312479        0.176145   0.086066         0.399476   
2     0.192106      0.240785        0.187478   0.097434         0.497156   
3     0.203464      0.124450        0.201852   0.102354         0.575697   
4     0.388518      0.118363        0.372193   0.241060         0.243748   

   compactness_mean  concavity_mean  points_mean  symmetry_mean  \
0          0.154684        0.093416     0.183897       0.454040   
1          0.292375        0.149649     0.131312       0.435354   
2          0.179928        0.071368     0.123260       0.330303   
3          0.289001        0.108599     0.238370       0.359091   
4          0.153242        0.079499     0.132058       0.333838   

   dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0        0.201980  ...      0.198150       0.096482         0.182081   
1        0.314869  ...      0.140519       0.291045         0.138802   
2        0.283067  ...      0.159374       0.384328         0.147019   
3        0.226622  ...      0.141942       0.099947         0.130086   
4        0.115417  ...      0.294201       0.098881         0.269386   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    0.089437          0.444628           0.096351         0.099201   
1    0.058887          0.331044           0.217530         0.153035   
2    0.070340          0.434062           0.117307         0.085224   
3    0.061148          0.432741           0.150294         0.069241   
4    0.155795          0.273592           0.142048         0.108786   

   points_worst  symmetry_worst  dimension_worst  
0      0.322715        0.248768         0.083104  
1      0.272371        0.271043         0.136626  
2      0.255361        0.282476         0.155910  
3      0.295911        0.105855         0.083956  
4      0.281031        0.181747         0.082776  

[5 rows x 30 columns]

In [917]:
data_new.describe()

radius_mean  texture_mean  perimeter_mean   area_mean  smoothness_mean  \
count   569.000000    569.000000      569.000000  569.000000       569.000000   
mean      0.338222      0.323965        0.332935    0.216920         0.394785   
std       0.166787      0.145453        0.167915    0.149274         0.126967   
min       0.000000      0.000000        0.000000    0.000000         0.000000   
25%       0.223342      0.218465        0.216847    0.117413         0.304595   
50%       0.302381      0.308759        0.293345    0.172895         0.390358   
75%       0.416442      0.408860        0.416765    0.271135         0.475490   
max       1.000000      1.000000        1.000000    1.000000         1.000000   

       compactness_mean  concavity_mean  points_mean  symmetry_mean  \
count        569.000000      569.000000   569.000000     569.000000   
mean           0.260601        0.208058     0.243137       0.379605   
std            0.161992        0.186785     0.192857       0.138456   
min            0.000000        0.000000     0.000000       0.000000   
25%            0.139685        0.069260     0.100944       0.282323   
50%            0.224679        0.144189     0.166501       0.369697   
75%            0.340531        0.306232     0.367793       0.453030   
max            1.000000        1.000000     1.000000       1.000000   

       dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count      569.000000  ...    569.000000     569.000000       569.000000   
mean         0.270379  ...      0.296663       0.363998         0.283138   
std          0.148702  ...      0.171940       0.163813         0.167352   
min          0.000000  ...      0.000000       0.000000         0.000000   
25%          0.163016  ...      0.180719       0.241471         0.167837   
50%          0.243892  ...      0.250445       0.356876         0.235320   
75%          0.340354  ...      0.386339       0.471748         0.373475   
max          1.000000  ...      1.000000       1.000000         1.000000   

       area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count  569.000000        569.000000         569.000000       569.000000   
mean     0.170906          0.404138           0.220212         0.217403   
std      0.139932          0.150779           0.152649         0.166633   
min      0.000000          0.000000           0.000000         0.000000   
25%      0.081130          0.300007           0.116337         0.091454   
50%      0.123206          0.397081           0.179110         0.181070   
75%      0.220901          0.494156           0.302520         0.305831   
max      1.000000          1.000000           1.000000         1.000000   

       points_worst  symmetry_worst  dimension_worst  
count    569.000000      569.000000       569.000000  
mean       0.393836        0.263307         0.189596  
std        0.225884        0.121954         0.118466  
min        0.000000        0.000000         0.000000  
25%        0.223127        0.185098         0.107700  
50%        0.343402        0.247782         0.163977  
75%        0.554639        0.318155         0.242949  
max        1.000000        1.000000         1.000000  

[8 rows x 30 columns]

In [913]:
data["diagnosis"].value_counts()

Benign       357
Malignant    212
Name: diagnosis, dtype: int64

In [918]:
data["diagnosis"] = data["diagnosis"].replace(["B", "M"],["Benign", "Malignant"])
data["diagnosis"]

0         Benign
1         Benign
2         Benign
3         Benign
4         Benign
         ...    
564       Benign
565       Benign
566    Malignant
567       Benign
568    Malignant
Name: diagnosis, Length: 569, dtype: object

In [919]:
data["diagnosis"]. value_counts()

Benign       357
Malignant    212
Name: diagnosis, dtype: int64

#### Randomization using 12345

In [921]:
import random
target = data["diagnosis"]
random.seed(12345)
indx = random.sample(range(0, 569), 569) 
data_new_rand = data_new.iloc[indx] 
target_rand = target.iloc[indx]

#### Splitting data into 80% Training and 20% Testing

In [922]:
list(data_new_rand)

['radius_mean',
 'texture_mean',
 'perimeter_mean',
 'area_mean',
 'smoothness_mean',
 'compactness_mean',
 'concavity_mean',
 'points_mean',
 'symmetry_mean',
 'dimension_mean',
 'radius_se',
 'texture_se',
 'perimeter_se',
 'area_se',
 'smoothness_se',
 'compactness_se',
 'concavity_se',
 'points_se',
 'symmetry_se',
 'dimension_se',
 'radius_worst',
 'texture_worst',
 'perimeter_worst',
 'area_worst',
 'smoothness_worst',
 'compactness_worst',
 'concavity_worst',
 'points_worst',
 'symmetry_worst',
 'dimension_worst']

In [923]:

from sklearn.model_selection import train_test_split

target = data["diagnosis"]

y = target_rand
x = data_new_rand  ## data_new_rand does not have the target variable as it was removed before normalizing.

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=12345)

### Using K Nearest Neighbors Algorithm for Classification|

In [924]:
from sklearn.neighbors import KNeighborsClassifier 
neigh = KNeighborsClassifier(n_neighbors=21) 
model = neigh.fit(x_train, y_train)

In [925]:
y_predict = model.predict(x_test)

/Users/pratik/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [926]:
print(confusion_matrix(y_test, y_predict)*100)

[[7400  100]
 [ 500 3400]]


In [927]:
print(accuracy_score(y_test,y_predict)*100)

94.73684210526315


#### The K-Nearest Neighbor classifier using k value of 21, gives us an accuracy of 94.74% in detecting the Breast Cancer.